mypy/checkstrformat.py

```diff
@@ -353,3 +902,45 @@ def named_type(self, name: str) -> Instance:
     def accept(self, expr: Expression, context: Optional[Type] = None) -> Type:
         """Type check a node. Alias for TypeChecker.accept."""
         return self.chk.expr_checker.accept(expr, context)
+
+
+def has_type_component(typ: Type, fullname: str) -> bool:
+    """Is this a specific instance type, or a union that contains it?
+
+    We use this ad-hoc function instead of a proper visitor or subtype check
+    because some str vs bytes errors are strictly speaking not runtime errors,
+    but rather highly counter-intuitive behavior. This is similar to what is used for
+    --strict-equality.
+    """
+    typ = get_proper_type(typ)
+    if isinstance(typ, Instance):
+        return typ.type.has_base(fullname)
+    elif isinstance(typ, TypeVarType):
+        return (has_type_component(typ.upper_bound, fullname) or
+                any(has_type_component(v, fullname) for v in typ.values))
+    elif isinstance(typ, UnionType):
+        return any(has_type_component(t, fullname) for t in typ.relevant_items())
+    return False
+
+
+def custom_special_method(typ: Type, name: str) -> bool:
```

Comment:
This will make it a runtime import cycle, currently 'checkstrformat' only imports others under `if TYPE_CHECKING: ...`. Maybe we should make a new module like `checker_utils.py`?